scala/collection/TraversableOnce.scala issues

```
dotc ./scala-scala/src/library/scala/collection/TraversableOnce.scala
./scala-scala/src/library/scala/collection/TraversableOnce.scala:437: error: class OnceCanBuildFrom needs to be abstract, since def apply:
  (from: [+hk$0] => scala.collection.TraversableOnce[hk$0])
    scala.collection.mutable.Builder[
      scala$collection$TraversableOnce$OnceCanBuildFrom$$A
    ,
      scala.collection.TraversableOnce[
        scala$collection$TraversableOnce$OnceCanBuildFrom$$A
      ]
    ] is not defined
(Note that TypeRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,generic)),CanBuildFrom)),scala$collection$generic$CanBuildFrom$$From) does not match RefinedType(TypeRef(ThisType(TypeRef(NoPrefix,collection)),TraversableOnce), scala$collection$TraversableOnce$$A, TypeBounds(TypeRef(ThisType(TypeRef(NoPrefix,scala)),Nothing), TypeRef(ThisType(TypeRef(NoPrefix,scala)),Any)) | 1651380528))
  class OnceCanBuildFrom[A] extends BufferedCanBuildFrom[A, TraversableOnce] {
          ^
one error found
```

--- a/src/dotty/tools/dotc/ast/Desugar.scala
+++ b/src/dotty/tools/dotc/ast/Desugar.scala
@@ -236,10 +236,12 @@ object desugar {
     // prefixed by type or val). `tparams` and `vparamss` are the type parameters that
     // go in `constr`, the constructor after desugaring.
 
+    val isCaseClass = mods.is(Case) && !mods.is(Module)
+
     val constrTparams = constr1.tparams map toDefParam
     val constrVparamss =
       if (constr1.vparamss.isEmpty) { // ensure parameter list is non-empty
-        if (mods is Case)
+        if (isCaseClass)
           ctx.error("case class needs to have at least one parameter list", cdef.pos)
         ListOfNil
       }
@@ -287,7 +289,7 @@ object desugar {
     // neg/t1843-variances.scala for a test case. The test would give
     // two errors without @uncheckedVariance, one of them spurious.
     val caseClassMeths =
-      if (mods is Case) {
+      if (isCaseClass) {
         def syntheticProperty(name: TermName, rhs: Tree) =
           DefDef(name, Nil, Nil, TypeTree(), rhs).withMods(synthetic)
         val isDefinedMeth = syntheticProperty(nme.isDefined, Literal(Constant(true)))
@@ -326,9 +328,9 @@ object desugar {
       if (targs.isEmpty) tycon else AppliedTypeTree(tycon, targs)
     }
 
-    // Case classes get a ProductN parent
+    // Case classes and case objects get a ProductN parent
     var parents1 = parents
-    if ((mods is Case) && arity <= Definitions.MaxTupleArity)
+    if (mods.is(Case) && arity <= Definitions.MaxTupleArity)
       parents1 = parents1 :+ productConstr(arity)
 
     // The thicket which is the desugared version of the companion object
@@ -350,7 +352,7 @@ object desugar {
     //     (T11, ..., T1N) => ... => (TM1, ..., TMN) => C
     // For all other classes, the parent is AnyRef.
     val companions =
-      if (mods is Case) {
+      if (isCaseClass) {
         val parent =
           if (constrTparams.nonEmpty ||
               constrVparamss.length > 1 ||
@@ -386,7 +388,7 @@ object desugar {
         ctx.error("implicit classes may not be toplevel", cdef.pos)
         Nil
       }
-      else if (mods is Case) {
+      else if (isCaseClass) {
         ctx.error("implicit classes may not be case classes", cdef.pos)
         Nil
       }
@@ -407,7 +409,7 @@ object desugar {
       val originalTparams = constr1.tparams.toIterator
       val originalVparams = constr1.vparamss.toIterator.flatten
       val tparamAccessors = derivedTparams.map(_.withMods(originalTparams.next.mods))
-      val caseAccessor = if (mods is Case) CaseAccessor else EmptyFlags
+      val caseAccessor = if (isCaseClass) CaseAccessor else EmptyFlags
       val vparamAccessors = derivedVparamss.flatten.map(_.withMods(originalVparams.next.mods | caseAccessor))
       cpy.TypeDef(cdef)(
         rhs = cpy.Template(impl)(constr, parents1, self1,
@@ -453,7 +455,7 @@ object desugar {
         .withPos(tmpl.self.pos orElse tmpl.pos.startPos)
       val clsTmpl = cpy.Template(tmpl)(self = clsSelf, body = tmpl.body)
       val cls = TypeDef(clsName, clsTmpl)
-        .withMods(mods.toTypeFlags & AccessOrSynthetic | ModuleClassCreationFlags)
+        .withMods(mods.toTypeFlags & RetainedModuleClassFlags | ModuleClassCreationFlags)
       Thicket(modul, classDef(cls))
     }
   }

--- a/src/dotty/tools/dotc/core/Definitions.scala
+++ b/src/dotty/tools/dotc/core/Definitions.scala
@@ -192,6 +192,9 @@ class Definitions {
     ScalaPackageClass, tpnme.Null, AbstractFinal, List(ObjectClass.typeRef))
 
   lazy val ScalaPredefModule = ctx.requiredModule("scala.Predef")
+
+    lazy val Predef_conforms = ctx.requiredMethod(ScalaPredefModule.moduleClass.asClass, "$conforms")
+
   lazy val ScalaRuntimeModule = ctx.requiredModule("scala.runtime.ScalaRunTime")
   lazy val ScalaRuntimeClass = ScalaRuntimeModule.moduleClass.asClass
 
@@ -210,7 +213,6 @@ class Definitions {
     def newRefArrayMethod = ctx.requiredMethod(DottyArraysModule.moduleClass.asClass, "newRefArray")
 
   lazy val NilModule = ctx.requiredModule("scala.collection.immutable.Nil")
-  lazy val PredefConformsClass = ctx.requiredClass("scala.Predef." + tpnme.Conforms)
 
 //  lazy val FunctionClass: ClassSymbol = ctx.requiredClass("scala.Function")
   lazy val SingletonClass: ClassSymbol =

--- a/src/dotty/tools/dotc/typer/Implicits.scala
+++ b/src/dotty/tools/dotc/typer/Implicits.scala
@@ -65,7 +65,9 @@ object Implicits {
           case tpw =>
             //if (ctx.typer.isApplicable(tp, argType :: Nil, resultType))
             //  println(i"??? $tp is applicable to $this / typeSymbol = ${tpw.typeSymbol}")
-            !tpw.derivesFrom(defn.FunctionClass(1)) || tpw.isRef(defn.PredefConformsClass)
+            !tpw.derivesFrom(defn.FunctionClass(1)) ||
+            ref.symbol == defn.Predef_conforms //
+              // as an implicit conversion, Predef.$conforms is a no-op, so exclude it
         }
 
         def discardForValueType(tpw: Type): Boolean = tpw match {

--- a/src/dotty/tools/dotc/typer/RefChecks.scala
+++ b/src/dotty/tools/dotc/typer/RefChecks.scala
@@ -287,7 +287,9 @@ object RefChecks {
                  !member.isAnyOverride) {
         // (*) Exclusion for default getters, fixes SI-5178. We cannot assign the Override flag to
         // the default getter: one default getter might sometimes override, sometimes not. Example in comment on ticket.
-        if (member.owner != clazz && other.owner != clazz && !(other.owner derivesFrom member.owner))
+        if (member.name == nme.isDefined && member.is(Synthetic)) // isDefined methods are added automatially, can't have an override preset.
+          member.setFlag(Override)
+        else if (member.owner != clazz && other.owner != clazz && !(other.owner derivesFrom member.owner))
           emitOverrideError(
             clazz + " inherits conflicting members:\n  "
               + infoStringWithLocation(other) + "  and\n  " + infoStringWithLocation(member)

--- a/test/dotc/scala-collections.whitelist
+++ b/test/dotc/scala-collections.whitelist
@@ -103,7 +103,7 @@
 ./scala-scala/src/library/scala/collection/LinearSeqOptimized.scala
 
 # https://github.com/lampepfl/dotty/issues/914
-#./scala-scala/src/library/scala/collection/TraversableOnce.scala
+./scala-scala/src/library/scala/collection/TraversableOnce.scala
 ./scala-scala/src/library/scala/collection/generic/Growable.scala
 ./scala-scala/src/library/scala/collection/generic/TraversableForwarder.scala
 ./scala-scala/src/library/scala/collection/immutable/BitSet.scala

--- a/test/dotc/scala-collections.whitelist
+++ b/test/dotc/scala-collections.whitelist
@@ -53,7 +53,7 @@
 ./scala-scala/src/library/scala/NotNull.scala
 
 # https://github.com/lampepfl/dotty/issues/911
-#./scala-scala/src/library/scala/Option.scala
+./scala-scala/src/library/scala/Option.scala
 
 ./scala-scala/src/library/scala/PartialFunction.scala
 ./scala-scala/src/library/scala/Predef.scala

--- a/tests/new/conforms.scala
+++ b/tests/new/conforms.scala
@@ -1,0 +1,3 @@
+object Test {
+  def f[A, B](x: A)(implicit e: <:<[A, B]): B = x
+}

